### PR TITLE
TOUR-81: Tour popover is not keyboard accessible

### PR DESCRIPTION
**Pull request by Sereza7**
**Jira:** https://jira.xwiki.org/browse/TOUR-81

## PR Changes:
* Added empty href to get the anchors in
* Moved the structure of the popup to avoid getting the last index
* Added event listeners on buttons to trap focus inside the tour when opened
* Added listeners to get keyboard event for space and return
* Added a listener to get keyboard escape event and leave the tour
* Fixed the popover structure, to avoid duplicate id generation
* Fixed style on the popup indication to restart the tour
* Added a fadein/fadeout animation to the popup indication to restart the tour
* Removed inline styling

## Demo video
https://up1.xwikisas.com/#2nUYrgbrOlMN0ld9r2kqJw

## Note
This would also fix [TOUR-69: Enable to close a Tour using ecape key](https://jira.xwiki.org/browse/TOUR-69)
## Tests
Built successfully the application module: `mvn clean install -Pquality`
As seen in the demo video, live tests of accessibility have been done, and there's no issue reported about the tour application anymore by axe-core.

### CODE REVIEW

**Comment by GridexX** [comment]
Just look at what you did. The demo video is nice, as I don't really know which modification are done, it is really cool to see in action your changes :+1: 

One remark, they are 7 identical commits maybe they should be squashed into a single one. It could mention both 81 and 69 fix ?

**Comment by Sereza7** [comment]
> Just look at what you did. The demo video is nice, as I don't really know which modification are done, it is really cool to see in action your changes +1
> 
> One remark, they are 7 identical commits maybe they should be squashed into a single one.

The commit first lines are identical, but not the message you can see when expanding them.
I don't squash before starting a merge request, I think it could be useful for reviewers, or even for me later.
And yeah, I don't mind my PRs getting squashed on merge :)

>  It could mention both 81 and 69 fix ?

A PR for xwiki platform should be linked to a ticket. Sometimes it solves multiple tickets, but we don't need to link every ticket to a unique PR.
I think I'll just close `TOUR-69` with the status `FixedBy` when this PR is merged (and put a comment on the ticket to explain).
Speaking of properly commenting, I just realized I forgot to link the issues on Jira earlier.

**Comment by GridexX** [comment]
@Sereza7 Thanks, for the explanation. Indeed, I didn't see the body of the commit

File: application-tour-ui/src/main/resources/TourCode/TourJS.xml [hunk 1]
```diff
@@ -249,16 +240,100 @@ require(['jquery', 'xwiki-meta'], function ($, xm) {
...
+      .focus();
 
-      $(closeButtonSelector).css('position','absolute');
-      $(closeButtonSelector).css('top','3px');
-      $(closeButtonSelector).css('right','5px');
-      // Avoid having the close button on top of the title
-      $('.tour .popover-title').css('padding-right',  $(closeButtonSelector).outerWidth() + 10 + 'px');
+    $(closeButtonSelector).css('position','absolute');
+    $(closeButtonSelector).css('top','3px');
+    $(closeButtonSelector).css('right','5px');
```
Comment (michitux) [inline application-tour-ui/src/main/resources/TourCode/TourJS.xml:325]: Can't you use some actual CSS for this?

File: application-tour-ui/src/main/resources/TourCode/StepSheet.xml [hunk 2]
```diff
@@ -564,6 +564,7 @@ if(typeof(XWiki) == 'undefined' || typeof(XWiki.widgets) == 'undefined' || typeo
    padding: .3em .3em .3em 20px;
    float: right;
 }
+
```
[the remark this answers is not in the document]
Comment (Sereza7) [reply application-tour-ui/src/main/resources/TourCode/StepSheet.xml]: Addressed in https://github.com/xwiki-contrib/application-tour/pull/26/commits/8ff702f11bbfd9f514af2425e7e0259c1b7e8e3c

File: application-tour-ui/src/main/resources/TourCode/TourJS.xml [hunk 3]
```diff
@@ -249,16 +240,100 @@ require(['jquery', 'xwiki-meta'], function ($, xm) {
...
+          tour.prev();
+        }
+        event.stopImmediatePropagation();
+        event.preventDefault();
+      })
+      .on('focusout', function (event) {
+        if (event.relatedTarget !== null &amp;&amp; !event.currentTarget.parentNode.parentNode.parentNode.contains(event.relatedTarget)) {
+          $(nextButtonSelector + ', ' + endButtonSelector).focus();
+          event.stopImmediatePropagation();
+          event.preventDefault();
```
[the remark this answers is not in the document]
Comment (Sereza7) [reply application-tour-ui/src/main/resources/TourCode/TourJS.xml]: 1. I'm limited by the use of the bootstrap tour library. This was already very limiting in changing the way the buttons are represented (see the oldest comment on https://jira.xwiki.org/browse/TOUR-81 ). This library allows to 'highlight' a part of the screen, that is not doable with a regular dialog AFAIK. 
2. Those stop propagation events are useless, I forgot to remove it, I added them out of habit and the changes work properly without them. Removed in https://github.com/xwiki-contrib/application-tour/pull/26/commits/a3b6918af68c1d45413376cda3529010bdaaf789.

File: application-tour-ui/src/main/resources/TourCode/TourJS.xml [hunk 4]
```diff
@@ -249,16 +240,100 @@ require(['jquery', 'xwiki-meta'], function ($, xm) {
...
+        event.preventDefault();
+      });
+
+    // Wrap around in tab order when the popover is opened, and focus the 'Next' button when the tour step is opened.
+    $(nextButtonSelector + ', ' + endButtonSelector)
+      .on('focusout', function (event) {
+        if (event.relatedTarget !== null &amp;&amp; !event.currentTarget.parentNode.parentNode.parentNode.contains(event.relatedTarget)) {
+          $(closeButtonSelector).focus();
+          event.stopImmediatePropagation();
+          event.preventDefault();
```
[the remark this answers is not in the document]
Comment (Sereza7) [reply application-tour-ui/src/main/resources/TourCode/TourJS.xml:314]: Addressed in https://github.com/xwiki-contrib/application-tour/pull/26/commits/a3b6918af68c1d45413376cda3529010bdaaf789. See https://github.com/xwiki-contrib/application-tour/pull/26#discussion_r1271958430

File: application-tour-ui/src/main/resources/TourCode/TourJS.xml (hunk 1 above)
Comment (Sereza7) [reply application-tour-ui/src/main/resources/TourCode/TourJS.xml]: This code was here before the PR. However before this PR there was no style extension object. We can easily add this CSS to the style sheet.
Addressed in https://github.com/xwiki-contrib/application-tour/pull/26/commits/d633e36bdbe4db0ebd8bcf3194da6c5dbec8f590.

**Comment by Sereza7** [comment]
Ready for reviews.
## PR Updates
* Improved code quality

File: application-tour-ui/src/main/resources/TourCode/TourJS.xml [hunk 5]
```diff
@@ -277,9 +345,26 @@ require(['jquery', 'xwiki-meta'], function ($, xm) {
       var tour     = new Tour({
         name    : tourName,
         storage : window.localStorage,
-        onEnd   : function() { createResumeButton(tour, true) },
+        onEnd   : function() {
+          let buttonContainer = $('#tourResumeContainer');
+          if(buttonContainer.length === 0) {
```
Comment (michitux) [inline application-tour-ui/src/main/resources/TourCode/TourJS.xml:350]:
```suggestion
          if (buttonContainer.length === 0) {
```

File: application-tour-ui/src/main/resources/TourCode/TourJS.xml [hunk 6]
```diff
@@ -277,9 +345,26 @@ require(['jquery', 'xwiki-meta'], function ($, xm) {
...
+            createResumeButton(tour, true);
+          }
+          else {
```
Comment (michitux) [inline application-tour-ui/src/main/resources/TourCode/TourJS.xml:353]:
```suggestion
          } else {
```

File: application-tour-ui/src/main/resources/TourCode/TourJS.xml [hunk 7]
```diff
@@ -356,4 +441,174 @@ require(['jquery', 'xwiki-meta'], function ($, xm) {
...
+    line-break: auto;
+    text-align: left;
+    padding:0.5em;
+    background-color: @dropdown-bg;
+    border: solid 0.2em @dropdown-border;
+    border-radius: 1em 1em 0 1em;
+    width: 100%;
+
+    &amp;::after {
+      position:absolute;
```
Comment (michitux) [inline application-tour-ui/src/main/resources/TourCode/TourJS.xml:572]:
```suggestion
      position: absolute;
```

File: application-tour-ui/src/main/resources/TourCode/TourJS.xml [hunk 8]
```diff
@@ -356,4 +441,174 @@ require(['jquery', 'xwiki-meta'], function ($, xm) {
...
+  &amp; &gt; .popover-content {
+    visibility: hidden;
+    opacity: 0;
+    transition: opacity 0.2s ease-in-out, visibility 0s 0.2s;
+    position: relative;
+    max-width: 276px;
+    padding: 1px;
+    line-break: auto;
+    text-align: left;
+    padding:0.5em;
```
Comment (michitux) [inline application-tour-ui/src/main/resources/TourCode/TourJS.xml:565]:
```suggestion
    padding: 0.5em;
```

File: application-tour-ui/src/main/resources/TourCode/TourJS.xml [hunk 9]
```diff
@@ -356,4 +441,174 @@ require(['jquery', 'xwiki-meta'], function ($, xm) {
...
+    background-color: @dropdown-bg;
+    border: solid 0.2em @dropdown-border;
+    border-radius: 1em 1em 0 1em;
+    width: 100%;
+
+    &amp;::after {
+      position:absolute;
+      right: 0em;
+      content: '';
+      bottom:-1em;
```
Comment (michitux) [inline application-tour-ui/src/main/resources/TourCode/TourJS.xml:575]:
```suggestion
      bottom: -1em;
```

**Review by michitux — APPROVED** [review]
Looks good to me. There are some small style fixes and I think you missed one of the `preventDefault` calls for `focusOut`.

**Comment by Sereza7** [comment]
Addressed all comments from https://github.com/xwiki-contrib/application-tour/pull/26#pullrequestreview-1617720986 :+1:

**Pull request merged**